### ISSUE TITLE
Cross Compile for Scala.js

### ISSUE THREAD

**Issue by mdedetrich** [cb372/scalacache#110]
The intention would essentially be to cross compile Scala-js for both the JVM and Scala.js (and maybe in the future, Scala Native!)

In terms of general design, this should be fairly easy (considering that we use macros). The main issue I see now is how the code is organised in terms of a project structure. I see a possible solution as follows

- scalacache-core becomes its own project and is cross compiled for scala-js/jvm
- The BaseCodecs (https://github.com/cb372/scalacache/blob/ee82acd128e3c9c984fe72e3d18b76c4f74c4266/core/src/main/scala/scalacache/serialization/BaseCodecs.scala) gets moved into its own project
    - Reason why is because there isn't really a corresponding byte type in Javascript. Although there is interopt with Byte (see https://www.scala-js.org/doc/interoperability/types.html) it will have terrible performance due to the semantics of Javascript treating it as an integer.
    - From a design POV basecodecs doesn't make too much sense being part of scalacore since the core shouldn't make assumptions on the type of serialization
    - Note that this isn't mandatory, its just that having the byte serializers can be considered deceiving when taking into account Javascript (for the reason stated earlier)
    - Some codecs (such as memcached) will depend off scalacache-basecodecs (pulling in the scalacache-core library transitively) rather than depending off scalacache-core directly. This ofcourse depends on the implementation of the scalacache module
- All of the scalacache modules are untouched (apart from changing their dependencies to point to the new scalacore/scalacore-basecodecs)
- Find a good implementation of a eviction based cache for Javascript, maybe something like https://github.com/monsur/jscache to start off with. Again this is not mandatory, but it would be nice to demonstrate a usable Scala.js cache.

Note that none of this should have any ramifications for end users, its mainly infrastructure changes.

Thoughts? I would also be willing to do the work if the author has no issues with it

**Comment by mdedetrich** [comment]
Also can hit #111 at the same time (build would have to be redone)

**Comment by cb372** [comment]
I think splitting the codecs out of the core could be tricky. I can't really envisage how that would work.

Can't we just make the Scala.js cache implementation use the `InMemoryRepr` codec, i.e. perform no serialization?

**Comment by mdedetrich** [comment]
I think so, I think we still have to split out a "core" regardless because the current caches are JVM only so they have to be compiled only for JVM.

I think for sure that the initial implementation should just use the `InMemoryRepr` codec although I can't remember accurately right now (going off the top of my head).

I might do some work in a new branch and see how far I get

**Comment by mdedetrich** [comment]
@cb372 This was actually a lot easier than suspected, I have done an initial commit and basically everything works. The current pull request is here https://github.com/cb372/scalacache/pull/115

Also please ignore the bfb0a5a/c09fb92/d558684 commits, my local git screwed up and there isn't any branch that references those commits (it should get cleaned up by githubs git gc)

**Comment by aappddeevv** [comment]
Any tests with scalajs? I would not mind using this with node-cache in some way, but I was wondering if the core, while compiling, has been tested somewhat.

**Comment by cb372** [comment]
@aappddeevv No, I've never actually used ScalaCache with scalajs. If you manage to integrate it with node-cache, a PR would be very welcome.

**Comment by LogicalTime** [comment]
+1 would love to use a caching library that works on both jvm and ScalaJS

**Comment by aappddeevv** [comment]
Seems to work.  Used it with node cache.  Had to create stub logging lib though.

**Comment by mdedetrich** [comment]
> Seems to work. Used it with node cache. Had to create stub logging lib though.

Yeah we really need a cross compiled logging library for this to happen though

**Comment by aappddeevv** [comment]
There are a couple but they are not widely used in the scala is world. Since logging is not heavily used in scalacsche, suggest abstracting it a bit.
